src/shogun/mathematics/linalg/Vector.cpp

```diff
@@ -0,0 +1,227 @@
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/Vector.h>
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+
+namespace shogun
+{
+
+template<class T>
+Vector<T>::Vector()
+{
+	init();
+}
+
+template<class T>
+Vector<T>::Vector(SGVector<T> const &vector)
+{
+	init();
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector + vector.vlen, m_data.get());
+	m_len = vector.vlen;
+}
+
+template<class T>
+Vector<T>::Vector(Vector<T> const &vector)
+{
+	init();
+
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.size())),
+									[](T* ptr) {SG_FREE(ptr);});
+	m_len = vector.m_len;
+
+
+	if (vector.onGPU())
+	{
+#ifdef HAVE_VIENNACL
+		m_onGPU = true;
+		m_gpu_impl = std::unique_ptr<GPUVectorImpl>(new GPUVectorImpl(*(vector.m_gpu_impl)));
+		transferToCPU();
+#endif
+	}
+	else
+	{
+		std::copy(vector.data(), vector.data() + vector.size(), m_data.get());
+	}
+}
+
+template<class T>
+Vector<T>::~Vector()
+{
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(SGVector<T> const &vector)
+{
+	m_data = std::shared_ptr<T>(reinterpret_cast<T*>(SG_MALLOC(aligned_t, vector.vlen)),
+									[](T* ptr) {SG_FREE(ptr);});
+	std::copy(vector.vector, vector.vector+vector.vlen, m_data.get());
```

Comment:
Ok some more thoughts, more explicit

 * *Definitely* no copying. It has to be possible that we pass two `SGVector`s to `linalg::dot` which causes an operation that is just like doing the dot-product with say eigen3
 * This means there is a problem with shared memory managment between `SGVector` (used to create this `Vector`) and this `Vector`. Deleting the `SGVector` might free memory and `Vector` causes a segfault. Dont want that. So we need to share the reference counting. This means that `Vector` *has* to be a subclass of `SGReferencedData`. 
 * Finally, there is a problem if data is transfered to GPU, because now, the original `SGVector` memory *can* safely be freed. So once we transfer data, reference counter can be decreased. This means transferring data "gives up" on data in main memory. Transferring back creates a *new* `SGVector` that is again reference counted. Maybe we should call the transfer in fact *move_to_gpu* (as original data might be freed)
* Need to think about alignment. Might have to be remove for now (and then later introduce when we align `SGVector` as well. I have limited understanding of this though.

My two cents. Let me know